File: hive/lib/test/runtime_mode_test.py
from __future__ import annotations

import unittest

from hive.lib.runtime_mode import detect_interactive_mode


class RuntimeModeTests(unittest.TestCase):
    def test_hive_headless_1_forces_headless(self) -> None:
        self.assertEqual(
            detect_interactive_mode({"HIVE_HEADLESS": "1"}),
            {"mode": "headless", "source": "env"},
        )

    def test_hive_headless_0_forces_interactive(self) -> None:
        self.assertEqual(
            detect_interactive_mode({"HIVE_HEADLESS": "0"}),
            {"mode": "interactive", "source": "env"},
        )

    def test_ci_true_with_no_override_resolves_headless(self) -> None:
        self.assertEqual(
            detect_interactive_mode({"CI": "true"}),
            {"mode": "headless", "source": "ci"},
        )

    def test_no_signals_resolves_interactive_default(self) -> None:
        self.assertEqual(
            detect_interactive_mode({}),
            {"mode": "interactive", "source": "default"},
        )

    def test_hive_headless_1_wins_over_ci_true(self) -> None:
        self.assertEqual(
            detect_interactive_mode({"HIVE_HEADLESS": "1", "CI": "true"}),
            {"mode": "headless", "source": "env"},
        )

    def test_hive_headless_0_wins_over_ci_true(self) -> None:
        self.assertEqual(
            detect_interactive_mode({"HIVE_HEADLESS": "0", "CI": "true"}),
            {"mode": "interactive", "source": "env"},
        )

    def test_unrecognized_ci_value_does_not_trigger_headless(self) -> None:
        self.assertEqual(
            detect_interactive_mode({"CI": "1"}),
            {"mode": "interactive", "source": "default"},
        )


if __name__ == "__main__":
    unittest.main()

File: hive/lib/test/runtime_mode.test.mjs
/**
 * runtime_mode.test.mjs
 *
 * Covers detectInteractiveMode()'s precedence tiers for story
 * hqp-1-runtime-mode-detection: env override > CI > default, plus the
 * HIVE_HEADLESS-wins-over-CI precedence case.
 */

import { describe, expect, it } from 'vitest';

import { detectInteractiveMode } from '../runtime_mode.js';

describe('detectInteractiveMode', () => {
  it('HIVE_HEADLESS=1 forces headless (source: env)', () => {
    expect(detectInteractiveMode({ env: { HIVE_HEADLESS: '1' } })).toEqual({
      mode: 'headless',
      source: 'env',
    });
  });

  it('HIVE_HEADLESS=0 forces interactive (source: env)', () => {
    expect(detectInteractiveMode({ env: { HIVE_HEADLESS: '0' } })).toEqual({
      mode: 'interactive',
      source: 'env',
    });
  });

  it('CI=true with no HIVE_HEADLESS resolves headless (source: ci)', () => {
    expect(detectInteractiveMode({ env: { CI: 'true' } })).toEqual({
      mode: 'headless',
      source: 'ci',
    });
  });

  it('no HIVE_HEADLESS and no CI resolves interactive (source: default)', () => {
    expect(detectInteractiveMode({ env: {} })).toEqual({
      mode: 'interactive',
      source: 'default',
    });
  });

  it('HIVE_HEADLESS=1 wins over CI=true (source: env)', () => {
    expect(
      detectInteractiveMode({ env: { HIVE_HEADLESS: '1', CI: 'true' } }),
    ).toEqual({ mode: 'headless', source: 'env' });
  });

  it('HIVE_HEADLESS=0 wins over CI=true (source: env)', () => {
    expect(
      detectInteractiveMode({ env: { HIVE_HEADLESS: '0', CI: 'true' } }),
    ).toEqual({ mode: 'interactive', source: 'env' });
  });

  it('unrecognized CI value (not the literal string "true") does not trigger headless', () => {
    expect(detectInteractiveMode({ env: { CI: '1' } })).toEqual({
      mode: 'interactive',
      source: 'default',
    });
  });
});
